### DISCUSSION TITLE
SD file -> WavDecoder -> FilterStream -> WavEncoder -> SD file?

### DISCUSSION THREAD

**Discussion by Sobhan-Nami** [pschatzmann/arduino-audio-tools#962] [Q&A] [answered]
Hi again. 
Thank you for your great library. it's been very helpful for my project and I love it. 

I want to read some wav files from SD and filter them by FIR filter and write back again. 
how can I read the wav file from sd, decode data, pass it to the filter, again encode them, and write to the file? 
I don't know how to pass Wav Decoder to Filtered Stream.

**Comment by pschatzmann** [comment, marked as answer]
I suggest that you read the documentation and have a look at the examples. 
I would copy from the SD to the decoder which connects to the filter which connects to the encoder which connects to the output file.
There is a comprehensive Wiki documentation for each topic.

Please also consider [these advices](https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working). Start simple and add one step after the other and test that it is doing what you expect.

**Reply by Sobhan-Nami** [reply]
I read all of your documents and examples.
my problem is EncodedAudioStream needs to have OutputStream so I can add FilteredStream as output but the FilteredStream constructor needs to have InputStream which is EncodedAudioStream. and the FilteredStream doesn't have output to connect the encoder to it.

**Comment by pschatzmann** [comment]
No, you think too complicated: the EncodedAudioStream has the FilteredStream as output and the FilteredStream has the EncodedAudioStream as output. You can just build a nice chain of the processing streams passing each output streams as argument.

**Reply by Sobhan-Nami** [reply]
so you say I can build something like this:

```
EncodedAudioStream wavEncoder(&des, new WAVEncoder());
FilteredStream<int16_t, double> inFiltered(wavEncoder, 1);
EncodedAudioStream wavDecoder(&inFiltered, new WAVDecoder()); 
StreamCopy fileCopier(wavDecoder, src);

 ```

**Reply by pschatzmann** [reply]
yes

**Reply by Sobhan-Nami** [reply]
Thank you for your support.
you are the best.
